File: examples/extension.rb
# Getting started with Equipe Extensions
#
# install required gems
# gem install sinatra jwt faraday faker rackup rerun
#
# The extension needs to be configured with this actions in the UI of Equipe
#
# | id | name  | path | role   | title      | target |
# |----|-------|------|--------|------------|--------|
# | 1  | modal |      | member | Modal demo | modal  |
# | 2  | forms |      | member | Forms demo | modal  |
#
# rerun "EXTENSION_SECRET=<SECRET> ruby extension.rb"

require 'sinatra'
require 'json'
require 'jwt'
require 'faraday'
require 'securerandom'

class Equipe
  attr_reader :token

  def initialize(token)
    @token = token
    @data, _ = JWT.decode token, ENV['EXTENSION_SECRET']
  end

  def style_url
    data.dig('payload', 'style_url')
  end

  def view_name
    data.dig('payload', 'name').to_sym
  end

  private
    attr_reader :data
end

set :protection, except: :frame_options

# Receive a POST request for webhooks from Equipe and print the payload
post '/' do
  puts JSON.parse(request.body.read)
  status 200
end

get '/' do
  @equipe = Equipe.new(params[:token])

  erb @equipe.view_name
end

__END__

@@ layout
<!DOCTYPE html>
<html>
  <head>
    <title>Equipe Extension</title>
    <link rel="stylesheet" href="<%= @equipe.style_url %>" media="all">
  </head>
  <body class="extension">
    <%= yield %>
  </body>
</html>

@@ modal
<code><%= JSON.pretty_generate(@data) %></code>

@@ forms
<form action="/action_page">
  <div class="form-group">
    <label for="email">Email address:</label>
    <input type="email" class="form-control" id="email">
  </div>
  <div class="form-group">
    <label for="pwd">Password:</label>
    <input type="password" class="form-control" id="pwd">
  </div>
  <div class="checkbox">
    <label><input type="checkbox"> Remember me</label>
  </div>
  <button type="submit" class="btn btn-default">Submit</button>
</form>

File: examples/extension.php
<?php
// allow cookie to be set in iframe
ini_set('session.cookie_samesite', 'None');
session_set_cookie_params(['samesite' => 'None', 'secure' => true]);
session_start();
//import jwt handler
require_once("/running/jwt/vendor/autoload.php");
  
use Firebase\JWT\JWT;
use Firebase\JWT\Key;

function getEquipeApi($url, $key, $debug)
{
	// Create a new cURL resource
	$ch = curl_init("https://app.equipe.com/meetings/".$url);

	$xkey = "x-api-key: $key";

	// Set the content type to application/json
	curl_setopt($ch, CURLOPT_HTTPHEADER, array($xkey));

	// Return response instead of outputting
	curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);

	// Execute the POST request
	$result = curl_exec($ch);
	$response = curl_getinfo($ch);
	// Close cURL resource
	curl_close($ch);
	switch ($response['http_code']) {
		case 200:  # OK
			$data = json_decode($result);
			break;
		default:
			$data = 'Unexpected HTTP code: ' . $response['http_code'];
			break;
	}
	if ($debug) {
		var_dump($data);
	}
	return $data;
}

#Check if Webhook
if ($_SERVER['REQUEST_METHOD'] === 'POST') {
     // The request is using the POST method
	 $decoded = json_decode(file_get_contents('php://input'));
} else if($_SERVER['REQUEST_METHOD'] === 'GET') {
  //get secret
  $key = $_ENV["EQUIPE_SECRET"];
  //get token
  $jwt = $_GET['token'];
  
  JWT::$leeway = 60; // $leeway in seconds
  $decoded = JWT::decode($jwt, new Key($key, 'HS256'));
}
//when modal or browser include basic HTML structure
if ($decoded->payload->target == "modal" || $decoded->payload->target == "browser") {
?>
  <html>
  	<head>
  		<link rel="stylesheet" href="<?php echo $decoded->payload->style_url ?>">
  		<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
  		<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/js/bootstrap.min.js"></script>
  	</head>

  <body class="extension">
  <?php
}
?>
	<?php
	if($decoded->event_name == "results") {
		// implement logic to handle result event
	} else if($decoded->event_name == "action" && $decoded->payload->target == "webhook") {
		switch($decoded->payload->name) {
			case "webhookname":
			  // implement logic to handle webhook
			default:
				break;
		}
	}	else {
  	switch ($decoded->payload->name) {
  		case "modalname":
			// save Equipe API Key in session to reuse when use makes new request
  			$_SESSION['EquipeKey'] = $decoded->api_key;
			$spliturl = explode("/", $decoded->payload->url);
			$showid = $spliturl[4];
			$compid = $spliturl[6];
			$showdata = getEquipeApi("$showid/settings.json", $decoded->api_key, false);
  			echo "<h2>Welkom by demo extension</h2>";
			echo "<p>Show: " . $showdata->meeting_name . "</p>";
  			break;
      default:
        break;
  	}
	}
if ($decoded->payload->target == "modal" || $decoded->payload->target == "browser") {
	
	?>
	
	</body>
</html>

  <?php
}
?>
